### PR TITLE
feature/mock-buckets

### PR DESCRIPTION
**Pull request by bwalsh**
## Summary
This PR adds targeted test coverage in `urlmanager` for two goals:

1. Strengthen provider-specific multipart signing assertions in fast unit-style tests.
2. Introduce an opt-in mock-server smoke test that validates `fake-gcs-server` and Azurite usage.

The result is a lightweight MVP that improves confidence in Azure/GCS behavior without making default test runs slow or environment-sensitive.

---

### 1) Fast confidence in multipart signing behavior
Developers need deterministic tests that verify URL signing output and multipart key conventions for Azure and GCS without cloud credentials or network calls.

### 2) Practical local mock-server validation
Issue context calls for a repeatable mock-testing approach. We need proof that:
- `fake-gcs-server` can be exercised from Go tests.
- Azurite can be exercised from Go tests.
- tests fail/skip gracefully in non-Docker environments.

---

## Test Architecture

This PR intentionally uses a two-layer MVP test strategy:

- **Layer A: Unit-style backend tests** (`urlmanager/multipart_backends_test.go`)
  - Directly test backend logic (`Init`, `SignPart`, URL/query contracts).
  - No Docker, no emulator dependency.
  - Fast and suitable for default `go test` loops.

- **Layer B: Opt-in mock-server smoke tests** (`urlmanager/mock_servers_test.go`)
  - Docker/Testcontainers-based checks against provider emulators.
  - Explicitly gated by `SYFON_E2E_MOCK_SERVERS=1`.
  - Skips in `testing.Short()` and skips cleanly when Docker is unavailable.

This balances speed and realism while staying MVP-sized.

---

### Azure multipart test expansion
- Expanded Azure multipart coverage from basic signing checks to:
  - `Init()` returns a non-empty upload ID.
  - `SignPart()` generates a valid URL with expected path and query parameters.
- Added/verified assertions for:
  - `comp=block`
  - presence of `blockid`
  - presence of `sig`
  - decoded block ID format: `<uploadID>:00000007`
  - expected escaped path segment behavior

### New GCS multipart happy-path test
- Added a lightweight test for `gcsMultipartBackend` that:
  - creates an in-memory sqlite DB and stores a GCS credential row,
  - generates RSA signing material in-test (service-account-style JSON),
  - calls `Init()` and `SignPart()`,
  - asserts multipart object path format:
    - `.syfon-multipart/<uploadID>/<key>/parts/<partNumber>`
  - validates key signed URL query parameters:
    - `X-Goog-Algorithm=GOOG4-RSA-SHA256`
    - `X-Goog-Credential` contains signer identity
    - `X-Goog-Signature` is present
    - `X-Goog-Expires` is positive and bounded by configured max

### Stability note
- GCS expiry assertions are intentionally tolerant to small timing drift (`<=900` instead of exact `900`) to avoid flaky checks caused by second-level timing differences.

---

### Added test
- `TestMVPMockServers_FakeGCSAndAzurite`

### fake-gcs-server subtest
- Starts Docker container `fsouza/fake-gcs-server:1.53.0` via Testcontainers.
- Configures GCS client with emulator endpoint and no auth.
- Performs bucket create + object write/read.
- Verifies round-trip payload integrity.

### Azurite subtest
- Starts Docker container `mcr.microsoft.com/azure-storage/azurite:3.35.0` via Testcontainers.
- Uses Azure shared-key credential and container client.
- Creates container + uploads/downloads blob.
- Verifies round-trip payload integrity.

### Execution gating and resilience
- Entire suite requires:
  - `SYFON_E2E_MOCK_SERVERS=1`
- Skips in:
  - `testing.Short()`
- Graceful skip if Docker is unavailable (actionable log message instead of hard failure).

---

### Default (fast path)
```bash
go test ./urlmanager -count=1
```

### Opt-in mock-server smoke tests
```bash
SYFON_E2E_MOCK_SERVERS=1 go test ./urlmanager -run TestMVPMockServers_FakeGCSAndAzurite -count=1 -v
```

---

## Validation Performed
- Ran focused tests for new multipart assertions.
- Ran focused test for `TestMVPMockServers_FakeGCSAndAzurite`.
- Ran full package test:
  - `go test ./urlmanager -count=1`
- Verified expected skip behavior for mock-server subtests when Docker is unavailable.

---

## Risks / Limitations
- Mock-server coverage is currently smoke-level (round-trip IO), not full Syfon server/CLI e2e.
- Emulator behavior may diverge from managed cloud services for advanced edge cases.
- Docker is required for emulator subtests; default test path remains Docker-free.

---

## Follow-ups (Optional)
1. Add dedicated CI job for `SYFON_E2E_MOCK_SERVERS=1` runs.
2. Extend emulator tests to cover multipart complete semantics.
3. Add higher-level Syfon server integration tests wired directly to emulator endpoints.

### CODE REVIEW

File: urlmanager/mock_servers_test.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+	"github.com/testcontainers/testcontainers-go/wait"
+	"google.golang.org/api/option"
+)
+
+const (
+	mockServersEnvVar  = "SYFON_E2E_MOCK_SERVERS"
+	fakeGCSImage       = "fsouza/fake-gcs-server:1.53.0"
+	azuriteImage       = "mcr.microsoft.com/azure-storage/azurite:3.35.0"
+	azuriteAccountName = "devstoreaccount1"
+	azuriteAccountKey  = "Eby8vdM02xNOcqFeqCnf2g=="
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply urlmanager/mock_servers_test.go]: done

File: urlmanager/mock_servers_test.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	if err == nil {
+		return false
+	}
+	lower := strings.ToLower(err.Error())
+	return strings.Contains(lower, "docker daemon") ||
+		strings.Contains(lower, "no such host") ||
+		strings.Contains(lower, "cannot connect") ||
+		strings.Contains(lower, "connection refused") ||
+		strings.Contains(lower, "failed to create docker provider") ||
+		strings.Contains(lower, "context deadline exceeded") ||
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply urlmanager/mock_servers_test.go]: done

File: urlmanager/multipart_backends_test.go [hunk 3]
```diff
@@ -96,3 +182,22 @@ func TestGCSSignedUploadPartURL_RequiresServiceAccountSigningMaterial(t *testing
 		t.Fatalf("unexpected error: %v", err)
 	}
 }
+
+func mustGCSServiceAccountJSON(t *testing.T, clientEmail string) string {
+	t.Helper()
+	privateKey, err := rsa.GenerateKey(rand.Reader, 1024)
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply urlmanager/multipart_backends_test.go]: done

File: urlmanager/mock_servers_test.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+		if err != nil {
+			t.Fatalf("marshal bucket create request: %v", err)
+		}
+		createBucketURL := endpointBase + "/storage/v1/b?project=syfon-mvp"
+		createReq, err := http.NewRequestWithContext(ctx, http.MethodPost, createBucketURL, bytes.NewReader(bucketBody))
+		if err != nil {
+			t.Fatalf("build create bucket request: %v", err)
+		}
+		createReq.Header.Set("Content-Type", "application/json")
+		createResp, err := http.DefaultClient.Do(createReq)
```
Comment (matthewpeterkort) [inline urlmanager/mock_servers_test.go:84]: Was going to say use the client function here for the bucket but I don't see one so I guess that is out of scope and add it to the todo list.

File: urlmanager/mock_servers_test.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+		if err != nil {
+			t.Fatalf("create bucket: %v", err)
+		}
+		if createResp.StatusCode < 200 || createResp.StatusCode > 299 {
+			_ = createResp.Body.Close()
+			t.Fatalf("unexpected create bucket status: %s", createResp.Status)
+		}
+		_ = createResp.Body.Close()
+
+		uploadURL := fmt.Sprintf("%s/upload/storage/v1/b/%s/o?uploadType=media&name=%s", endpointBase, bucketName, url.QueryEscape(objectName))
```
Comment (matthewpeterkort) [inline urlmanager/mock_servers_test.go:94]: ok but this function should have a client function. Same with non bucket endpoints below. It order for these tests to be most effective the should attempt to leverage the client code the syfon server code using these mock data providers

**Review by matthewpeterkort — CHANGES_REQUESTED** [review]

File: urlmanager/mock_servers_test.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+	})
+}
+
+func isDockerUnavailableForMockTests(err error) bool {
+	if err == nil {
+		return false
+	}
+	lower := strings.ToLower(err.Error())
+	return strings.Contains(lower, "docker daemon") ||
+		strings.Contains(lower, "no such host") ||
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply urlmanager/mock_servers_test.go]: done

**Comment by bwalsh** [comment]
Added client side tests


## PR Description: Multi-Storage Client/Server Test Architecture and Implementation

### Summary
This PR expands Syfon's multi-cloud confidence by adding targeted tests across both server and client paths for Google Cloud Storage (GCS) and Azure Blob Storage semantics. The approach is intentionally layered: deterministic unit tests for URL/signing behavior, integration-style client tests for end-to-end resolver flow, and Docker-backed emulator smoke tests for realistic provider interactions.

Together, these tests improve assurance that **Syfon server URL/signing outputs** and **Syfon client resolution/canonicalization logic** remain compatible across multiple storage backends.

---

### Architecture

This change follows a three-layer architecture:

1. **Server multipart/signing unit layer**
   - `urlmanager/multipart_backends_test.go`
   - Validates provider-specific multipart behavior without requiring cloud infrastructure.
   - Focuses on Azure/GCS signing contracts and multipart key semantics.

2. **Server emulator smoke layer (Docker/Testcontainers)**
   - `urlmanager/mock_servers_test.go`
   - Validates that fake-gcs-server and Azurite can be started and exercised in CI.
   - Uses opt-in gating (`SYFON_E2E_MOCK_SERVERS=1`) and fail-fast behavior for non-environmental startup regressions.

3. **Client provider-resolution layer**
   - `client/drs/access_method_conversions_test.go`
   - `client/drs/client_storage_location_test.go`
   - `client/drs/resolve_download_url_integration_test.go`
   - `client/data_service_test.go`
   - Validates client interpretation and preservation of `gs://` and `azblob://` URLs when consuming server metadata.

---

### Implementation Details

#### 1) Client access-method conversion coverage
- Added tests in `client/drs/access_method_conversions_test.go`.
- Verifies `DRSAccessMethodsFromInternalURLs` correctly derives access method type from URL scheme:
  - `gs://...` -> `gs`
  - `azblob://...` -> `azblob`
- Confirms authorization metadata is preserved through conversion.

#### 2) Client storage location extraction across providers
- Added tests in `client/drs/client_storage_location_test.go`.
- Updated `DrsClient.GetStorageLocation` in `client/drs/client.go` to parse:
  - `s3://bucket/key`
  - `gs://bucket/key`
  - `azblob://container/key`
- Added invalid-case assertion for malformed storage URLs.
- Result: bucket/key extraction is provider-aware while preserving backward-compatible fallback parsing.

#### 3) Integration-style client resolver test
- Added `TestDrsClientResolveDownloadURL_DirectAccessURL_GCSAndAzure` in `client/drs/resolve_download_url_integration_test.go`.
- Mocks Syfon object metadata response from `/index/{id}` containing direct provider URLs.
- Exercises `ResolveDownloadURL` end-to-end and asserts:
  - metadata path is called (`/index/{id}`),
  - direct URL is returned unchanged,
  - `/ga4gh/drs/v1/objects/{id}/access/{access_id}` is not used when a direct access URL is present.
- Covers both:
  - `gs://gcs-bucket/path/to/object.bin`
  - `azblob://az-container/path/to/object.bin`

#### 4) Client canonicalization behavior for GCS and Azure
- Added tests in `client/data_service_test.go`.
- Verifies `DataService.CanonicalObjectURL` behavior for:
  - HTTPS signed provider URLs -> normalized internal `s3://bucket/key` form (with bucket hint),
  - native `gs://...` and `azblob://...` URLs -> preserved as-is.
- This keeps internal object identity deterministic while respecting explicit provider-native URLs.

#### 5) Server emulator reliability and CI signal quality
- `urlmanager/mock_servers_test.go` uses:
  - fake-gcs-server JSON API round-trip,
  - Azurite blob upload/download with correct default `devstoreaccount1` key.
- Docker-unavailable matcher was tightened to avoid masking real regressions:
  - generic DNS/image-pull failures (e.g., registry `no such host`) now fail tests,
  - true daemon/connectivity issues still skip with explicit messaging.

---

### How this ensures client/server compatibility across multiple storage servers

These tests validate both sides of the contract:

- **Server side:** generates and handles provider-specific multipart and signing behavior for GCS/Azure.
- **Client side:** correctly interprets DRS object metadata and resolves provider URLs (`gs://`, `azblob://`) without accidental S3-only assumptions.
- **Integration bridge:** direct object metadata emitted by server APIs is consumable by client resolver logic without requiring provider-specific special casing at call sites.

This substantially reduces risk of regressions where one side changes and silently breaks multi-provider workflows.

---

### CI and Coverage Notes

- Mock-server tests are now wired into CI execution with `SYFON_E2E_MOCK_SERVERS=1`.
- Coverage run is executed with the same env flag so these code paths are exercised during coverage collection.
- Added matcher test coverage (`TestIsDockerUnavailableForMockTests`) to lock in skip/fail semantics.

---

### Validation Performed

- Focused client test runs for new GCS/Azure coverage.
- Focused integration-style resolver test run.
- Full `client` module test run (`go test ./...`).
- Docker-backed mock-server test run for fake-gcs-server and Azurite.

All targeted runs passed locally.

**Comment by bwalsh** [comment]
## Summary
This PR adds Docker-backed end-to-end coverage for GCS and Azure emulators, then updates URL/canonicalization behavior so the full Syfon CLI upload/download flow works reliably against those backends.

Specifically, it introduces:
- new fake-gcs-server and Azurite E2E suites under `cmd/`
- shared E2E helper logic to exercise all core CLI commands against a real Syfon process
- emulator compatibility fixes in URL signing, upload behavior, and canonical URL derivation
- docs updates for local execution
- CI updates to ensure these new suites run on every CI run

## Why
We already had Docker-backed MinIO coverage, but lacked equivalent validation for:
- GCS-style flows (fake-gcs-server)
- Azure Blob flows (Azurite)

Without these tests, provider-specific URL formatting/signing/canonicalization regressions can slip through even when unit tests and MinIO coverage pass.

## What changed

### 1) New Docker E2E suites
Added:
- `cmd/docker_fakegcs_e2e_test.go`
- `cmd/docker_azurite_e2e_test.go`
- `cmd/docker_cloud_e2e_helpers_test.go`

These suites:
- start provider emulator containers (fake-gcs-server / Azurite)
- start a real Syfon server process configured for that provider
- run the CLI against that server and exercise:
  - `version`
  - `ping`
  - `upload`
  - `ls`
  - `download`
  - `sha256sum`
  - `add-url`
  - `rm`
  - `bucket add/list/remove`

### 2) GCS emulator compatibility fixes
Updated:
- `urlmanager/gcs.go`
- `client/xfer/http.go`
- `client/data_service.go`

Key fixes:
- generate fake-gcs-compatible HTTP API URLs when a GCS endpoint is configured
  - upload path uses `/upload/storage/v1/b/...` with `uploadType=media`
  - download path uses `/storage/v1/b/.../o/...` with `alt=media`
- use `POST` for GCS JSON media uploads (instead of always `PUT`)
- canonicalize fake-gcs upload URLs back to correct internal object URL shape (bucket/key)

### 3) Azure/Azurite compatibility fixes
Updated:
- `urlmanager/azure.go`
- `urlmanager/manager.go`
- `client/xfer/http.go`
- `client/data_service.go`

Key fixes:
- support HTTP SAS protocol for local endpoints (Azurite)
- add `x-ms-blob-type: BlockBlob` header for signed blob PUT uploads
- prefer explicit Azure shared-key signing path for Azure provider URL signing
- improve canonical URL parsing for Azure/Azurite signed URLs so container/key are preserved correctly

### 4) Unit test coverage for canonicalization
Updated:
- `client/data_service_test.go`

Added cases for:
- fake-gcs JSON upload URL canonicalization
- overriding incorrect bucket hints for fake-gcs URLs
- Azure signed URL canonicalization
- Azurite signed URL canonicalization

### 5) Docs and CI
Updated:
- `README.md`
- `.github/workflows/ci.yaml`

Docs:
- added short instructions to run the new Docker E2E suites locally

CI:
- integration job now explicitly runs:
  - `TestSyfonDockerMinIOE2E`
  - `TestSyfonDockerMultipartUpload`
  - `TestSyfonDockerFakeGCSE2E`
  - `TestSyfonDockerAzuriteE2E`

## Validation performed

### Local targeted tests
```bash
cd ~/calypr/syfon
SYFON_E2E_DOCKER=1 go test ./cmd -run '^TestSyfonDockerFakeGCSE2E$' -count=1 -v
SYFON_E2E_DOCKER=1 go test ./cmd -run '^TestSyfonDockerAzuriteE2E$' -count=1 -v
```

### Additional compile/unit checks during iteration
```bash
cd ~/calypr/syfon/client
go test ./... -run '^TestDataServiceCanonicalObjectURL_GCSAndAzure$' -count=1

cd ~/calypr/syfon
go test ./urlmanager -run TestNonExistent -count=1
```

## Risk / impact
- Scope is primarily test harness + provider compatibility behavior for emulator flows.
- Production cloud behavior should remain unchanged; most special handling is gated by endpoint shape/signature form or provider type.
- Main risk area is canonical URL derivation for HTTP signed URLs; this PR adds unit tests to reduce regression risk.

## Notes
- Docker-backed suites remain opt-in locally via `SYFON_E2E_DOCKER=1`.
- CI now runs these suites automatically in the Docker integration job.

## Coverage delta (this branch)
- `urlmanager` package coverage improved from **51.6%** to **59.6%**.
- `client/xfer` package coverage improved from **13.3%** to **42.2%**.

Key function-level improvements:
- `urlmanager/manager.go`
  - `SignDownloadPart`: **0.0% -> 52.0%**
  - `CompleteMultipartUpload`: **0.0% -> 71.4%**
  - `isSigningNotSupported`: **75.0% -> 100.0%**
- `urlmanager/azure.go`
  - `azureAccountFromEndpoint`: **0.0% -> 80.0%**
- `client/xfer/http.go`
  - `DoUpload`: **0.0% -> 84.4%**
  - `GenericDownload`: **0.0% -> 100.0%**
  - `needsAzureBlobTypeHeader`: **100.0%**
  - `useGCSJSONMediaUpload`: **85.7%**

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -135,12 +137,56 @@
...
+          files: coverage/cmd-docker.out
+          flags: root
+          disable_search: true
+          verbose: true
+          fail_ci_if_error: true
+
+      - name: Upload Docker E2E cmd coverage to Codecov (token fallback)
+        if: steps.codecov_docker_cmd_oidc.outcome == 'failure'
+        continue-on-error: true
+        uses: codecov/codecov-action@v5
```
[the remark this answers is not in the document]
Comment (bwalsh) [reply .github/workflows/ci.yaml:172]: wont fix

**Comment by lbeckman314** [comment]
# Review Steps 🌀

## 1. Update

```sh
➜ cd syfon

➜ git checkout feature/mock-buckets

➜ git show --summary --oneline
4cb902a (HEAD -> feature/mock-buckets, origin/feature/mock-buckets) coverage tests
```

## 2. Run Tests

> [!TIP]
>
> Need to set `CGO_ENABLED=1` here for [go-sqlite3](https://github.com/mattn/go-sqlite3/tree/v1.14.42#compilation) support

```sh
➜ CGO_ENABLED=1 go test ./urlmanager -count=1
ok      github.com/calypr/syfon/urlmanager      3.132s

➜ SYFON_E2E_MOCK_SERVERS=1 go test ./urlmanager -run TestMVPMockServers_FakeGCSAndAzurite -count=1 -v

--- PASS: TestMVPMockServers_FakeGCSAndAzurite (42.61s)
    --- PASS: TestMVPMockServers_FakeGCSAndAzurite/fake-gcs-server (4.95s)
    --- PASS: TestMVPMockServers_FakeGCSAndAzurite/azurite (37.65s)
PASS
ok      github.com/calypr/syfon/urlmanager      43.861s
```

## 3. Run Server + Mock Buckets

> [!CAUTION]
>
> Work in Progress... 🚧

```sh
➜ make install

➜ which syfon
$HOME/go/bin/syfon

➜ syfon version
Syfon — a Modern DRS Server ⚡️
git commit:   4cb902a
git branch:   feature/mock-buckets
git upstream: origin/feature/mock-buckets
build date:   2026-04-16T00:16:06Z
version:      v0.1.11-13-g4cb902a-dirty

➜ syfon serve --config local.yaml
```

File: urlmanager/mock_servers_test.go (hunk 5 above)
Comment (bwalsh) [reply urlmanager/mock_servers_test.go:94]: @matthewpeterkort See tests added below in additional commits eg client/

**Review by matthewpeterkort — APPROVED** [review]
